wavesrv/pkg/blockstore/blockstore.go

```diff
@@ -0,0 +1,389 @@
+package blockstore
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/alecthomas/units"
+)
+
+type FileOptsType struct {
+	MaxSize  int64
+	Circular bool
+	IJson    bool
+}
+
+type FileMeta = map[string]any
+
+type FileInfo struct {
+	BlockId   string
+	Name      string
+	Size      int64
+	CreatedTs int64
+	ModTs     int64
+	Opts      FileOptsType
+	Meta      FileMeta
+}
+
+const MaxBlockSize = int64(128 * units.Kilobyte)
+
+type CacheEntry struct {
+	Lock       *sync.Mutex
+	CacheTs    int64
+	Info       FileInfo
+	DataBlocks []*CacheBlock
+}
+
+type CacheBlock struct {
+	data []byte
+	size int
+}
+
+func MakeCacheEntry(info FileInfo) *CacheEntry {
+	rtn := &CacheEntry{Lock: &sync.Mutex{}, CacheTs: int64(time.Now().UnixMilli()), Info: info, DataBlocks: []*CacheBlock{}}
+	return rtn
+}
+
+// add ctx context.Context to all these methods
+type BlockStore interface {
+	MakeFile(ctx context.Context, blockId string, name string, meta FileMeta, opts FileOptsType) error
+	WriteFile(ctx context.Context, blockId string, name string, meta FileMeta, data []byte) error
+	AppendData(ctx context.Context, blockId string, name string, p []byte) error
+	WriteAt(ctx context.Context, blockId string, name string, p []byte, off int64) (int, error)
+	ReadAt(ctx context.Context, blockId string, name string, p *[]byte, off int64) (int, error)
+	Stat(ctx context.Context, blockId string, name string) (FileInfo, error)
+	CollapseIJson(ctx context.Context, blockId string, name string) error
+	WriteMeta(ctx context.Context, blockId string, name string, meta FileMeta) error
+	DeleteFile(ctx context.Context, blockId string, name string) error
+	DeleteBlock(ctx context.Context, blockId string) error
+	ListFiles(ctx context.Context, blockId string) []FileInfo
+	FlushCache(ctx context.Context) error
+	GetAllBlockIds(ctx context.Context) []string
+}
+
+var cache map[string]*CacheEntry = make(map[string]*CacheEntry)
```

Comment:
currently I use global memory like in sstore.go, but since BlockStore is an interface I could make a `BlockStoreType` with all of the data. Then we could have multiple block stores, like if we wanted a different one for web communication or something like that that might be a use case. Should I do that?